README.md
<p align="center">
  <a href="http://nestjs.com/" target="blank">
    <img src="https://nestjs.com/img/logo_text.svg" width="320" alt="Nest Logo" />
  </a>
</p>

<p align="center">
  A short message module for Nest framework (node.js) using <a href="https://www.aliyun.com/product/sms">Alicloud SMS</a> service.   
</p>

<p align="center">
  <a href="https://www.npmjs.com/package/nestjs-alicloud-sms"><img src="https://img.shields.io/npm/v/nestjs-alicloud-sms.svg" alt="NPM Version" /></a>
  <a href="https://www.npmjs.com/package/nestjs-alicloud-sms"><img src="https://img.shields.io/npm/l/nestjs-alicloud-sms.svg" alt="Package License" /></a>
</p>

English | [简体中文](README-zh_CN.md)

## Before Installation

1. Create Alicloud account
2. Purchase SMS. ([read more](https://www.aliyun.com/product/sms))
3. Apply for SMS sign name and message template, go to **Dashboard > Short Message Service > Domestic Message > Add Singname**
4. Create message template, go to **Dashboard > Short Message Service > Domestic Message > Templates Management > Add Template**.

### Installation

```bash
npm install --save nestjs-alicloud-sms
#or
yarn add nestjs-alicloud-sms
```

### Documentation

1. Import the `AlicloudSmsModule` with the following configuration:

```typescript
import { Module } from '@nestjs/common';
import { AlicloudSmsModule } from 'nestjs-alicloud-sms';

import { AppController } from './app.controller';
import { AppService } from './app.service';

@Module({
  controllers: [AppController],
  providers: [AppService],
  imports: [
    AlicloudSmsModule.forRoot({
      config: { accessKeyId: '***', accessKeySecret: '***' },
      defaults: { signName: '***', regionId: 'cn-hangzhou' },
    }),
  ],
})
export class AppModule {}
```

2. Send Message

```typescript
import { Controller, Logger, Post } from '@nestjs/common';
import { AlicloudSmsService } from 'nestjs-alicloud-sms';

@Controller()
export class AppController {
  constructor(private readonly smsService: AlicloudSmsService) {}

  @Post('sms/send')
  async sendSms(): Promise<any> {
    const templateCodeId = 'SMS_1490999999';
    const phoneNumber = '1388886666';
    const templateParam = { code: '888666' };

    const sendResponse = await this.smsService.sendSms(templateCodeId, phoneNumber, templateParam);

    if (sendResponse.Code === 'OK') {
      // success
    } else {
      // failed
    }
  }
}
```

### License

The [MIT](LICENSE) License.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/YuchenWell/nestjs-alicloud-sms/compare/v1.0.0...v1.0.1) (2020-06-26)

## 1.0.0 (2020-06-24)

## 1.0.0 (2020-06-24)


lib/alicloud-sms.service.spec.ts
jest.mock('@alicloud/pop-core');
import * as RPCClient from '@alicloud/pop-core';

import { AlicloudSmsService } from './alicloud-sms.service';
import { AlicloudSmsOptions } from './interfaces';

RPCClient.prototype.request = (...args: any): any => {
  return {
    Code: 'OK',
  };
};

describe('AlicloudSmsService (without default options)', () => {
  let smsService: AlicloudSmsService = null;
  let options: AlicloudSmsOptions;

  beforeEach(() => {
    options = {
      config: {
        accessKeyId: '****',
        accessKeySecret: '****',
      },
    };

    smsService = new AlicloudSmsService(options);
  });

  it('should throw error if config is invalid that signname is not provided', async () => {
    await expect(smsService.sendSms('**', '**', {})).rejects.toThrow();
  });

  it('should throw error if config is invalid that regionId is not provided', async () => {
    await expect(smsService.sendSms('**', '**', {})).rejects.toThrow();
  });

  it('should send message successfully if provide signname and regionId via sendSms function', async () => {
    await expect(smsService.sendSms('**', '**', {}, 'signname', 'reginId')).resolves.toEqual({
      Code: 'OK',
    });
  });
});

describe('AlicloudSmsService (with default options)', () => {
  let smsService: AlicloudSmsService = null;
  let options: AlicloudSmsOptions;

  beforeEach(() => {
    options = {
      config: {
        accessKeyId: '****',
        accessKeySecret: '****',
      },
      defaults: {
        regionId: '****',
        signName: '****',
      },
    };

    smsService = new AlicloudSmsService(options);
  });

  it('should send message successfully', async () => {
    await expect(smsService.sendSms('**', '**', {})).resolves.toEqual({
      Code: 'OK',
    });
  });
});


lib/interfaces/alicloud-sms-options.interface.ts
export interface AlicloudSmsOptions {
  config: {
    /**
     * Alicloud account accessKey ID.
     *
     * @see https://usercenter.console.aliyun.com/#/manage/ak
     * @type {string}
     */
    accessKeyId: string;

    /**
     * Alicloud account accessKey secret.
     *
     * @see https://usercenter.console.aliyun.com/#/manage/ak
     * @type {string}
     */
    accessKeySecret: string;

    /**
     * Alicloud API service url.
     *
     * @default 'https://dysmsapi.aliyuncs.com'
     * @type {string}
     */
    endpoint?: string;

    /**
     * Alicloud SMS API version.
     *
     * @default 2017-05-25
     * @type {string}
     */
    apiVersion?: string;

    opts?: {
      /**
       * @default 3000
       * @type {number}
       */
      timeout?: number;

      /**
       * Format the parameter name to first letter upper case
       *
       * @default true
       * @type {boolean}
       */
      formatParams?: boolean;

      /**
       * Set the http method
       *
       * @default GET
       * @type {('GET' | 'POST')}
       */
      method?: 'GET' | 'POST';

      /**
       * Http request headers
       *
       * @type {object}
       */
      headers?: object;
    };
  };
  defaults?: {
    /**
     * SMS message template ID.
     *
     * @see 请在控制台模板管理页面模板CODE一列查看。
     * @type {string}
     */
    signName?: string;

    /**
     * Alicloud region ID.
     *
     * @type {string}
     */
    regionId?: string;
  };

  /**
   * Log sent message on dashboard.
   *
   * @default false
   * @type {boolean}
   * @memberof AlicloudSmsOptions
   */
  logger?: boolean;
}


lib/interfaces/alicloud-sms-response.interface.ts
export interface AlicloudSmsResponse {
  /**
   * Request status code.
   * - Return ```OK``` means the message was sent successfully.
   * - Other errors see [error code list](https://help.aliyun.com/document_detail/101346.htm)
   *
   * @type {string}
   * @memberof AlicloudSmsResponse
   */
  Code: string;

  /**
   * Status code description.
   * - Return ```OK``` means the message was sent successfully.
   * - Other errors see [error code list](https://help.aliyun.com/document_detail/101346.htm)
   *
   * @type {string}
   * @memberof AlicloudSmsResponse
   */
  Message: string;

  /**
   * Send receipt ID.
   * ONLY the message was sent successfully, this field will be returned.
   *
   * @type {string}
   * @memberof AlicloudSmsResponse
   */
  BizId?: string;

  /**
   * Request ID.
   *
   * @type {string}
   * @memberof AlicloudSmsResponse
   */
  RequestId?: string;
}


lib/alicloud-sms.service.ts
import { Inject, Injectable, Logger } from '@nestjs/common';

import * as RPCClient from '@alicloud/pop-core';

import { ALICLOUD_SMS_MODULE_OPTIONS } from './alicloud-sms.constant';
import { AlicloudSmsOptions } from './interfaces/alicloud-sms-options.interface';
import { AlicloudSmsResponse } from './interfaces/alicloud-sms-response.interface';

@Injectable()
export class AlicloudSmsService {
  private client: RPCClient;

  constructor(
    @Inject(ALICLOUD_SMS_MODULE_OPTIONS)
    private readonly options: AlicloudSmsOptions,
  ) {
    const config = {
      ...{
        endpoint: 'https://dysmsapi.aliyuncs.com',
        apiVersion: '2017-05-25',
      },
      ...options.config,
    };

    this.client = new RPCClient(config);
  }

  /**
   * Send message.
   * @template T
   * @param {string} templateCode 短信模板ID
   * @param {string | string[]} phoneNumber 接收短信的手机号码
   * @param {T} templateParam 短信模板变量对应的实际值，JSON格式
   * @param {string} [signName] 短信签名名称
   * @param {string} [regionId] 地区ID
   * @returns {(Promise<boolean | null>)}
   * @memberof AlicloudSmsService
   */
  public async sendSms(
    templateCode: string,
    phoneNumbers: string | string[],
    templateParam: object | string,
    signName?: string,
    regionId?: string,
  ): Promise<AlicloudSmsResponse | null> {
    if (!signName && !this.options?.defaults?.signName) {
      Logger.error('Error encountered: "SignName" was not provided.');
    }

    if (!regionId && !this.options?.defaults?.regionId) {
      Logger.error('Error encountered: "RegionId" was not provided.');
    }

    const params = {
      RegionId: regionId ?? this.options.defaults.regionId,
      SignName: signName ?? this.options.defaults.signName,
      PhoneNumbers: Array.isArray(phoneNumbers) ? phoneNumbers.join(',') : phoneNumbers,
      TemplateCode: templateCode,
      TemplateParam: typeof templateParam === 'string' ? templateParam : JSON.stringify(templateParam),
    };

    try {
      const requestOption = { method: 'POST' };

      const response: AlicloudSmsResponse = await this.client.request('SendSms', params, requestOption);

      if (this.options.logger) {
        if (response.Message === 'OK') {
          Logger.log(`Sent message to "${params.PhoneNumbers}" successfully.`, 'AlicloudSmsModule');
        } else {
          Logger.warn(
            `Sent message to "${params.PhoneNumbers}" failed, response "${response.Message}".`,
            'AlicloudSmsModule',
          );
        }
      }

      return response;
    } catch (err) {
      throw new Error(err);
    }
  }
}
